implementation.md

```diff
@@ -217,41 +215,127 @@ instance, storing it as a JSON file, then using that on another instance.
 method. Since there is no user-defined error-handling, these functions should
 not throw. Also, since Shesmu has no null values, they should not return null
 when an error occurs.
 
-To provide functions:
-
-1. Create a class that implements `FunctionRepository`.
-1. Annotate this class with `@MetaInfServices`.
-1. Include a no-arguments constructor for the `FunctionRepository` class.
-1. Return a stream of `FunctionDefintion` objects for the functions to be created.
-
-There are a few ways methods to generate a `FunctionDefinition`:
+##### File-Backed Matched Implementation
+For file-backed matched implementations, functions can be defined in three ways:
+
+- as static members of _R_
+- as static members of _R_ with the first argument being _T_
+- as instance members of _T_ (_e.g._, [`activeProjects()` in `PinerySource`](plugin/niassa+pinery/src/main/java/ca/on/oicr/gsi/shesmu/pinery/PinerySource.java))
+
+In all cases, the method must be public and annotated with `@ShesmuMethod`. The
+method must either have a Shesmu-compatible name or the `name` property of the
+annotation must be set to a Shesmu compatible name. For methods taking an
+instance of _T_, the name must include `$`, which will be substituted for the
+name of the configuration file.
+
+The `@ShesmuMethod` annotation also has a `description` property that will be
+shown in the definition page. In the description, `{instance}` and `{file}`
+will be replace with the instance name and configuration file path,
+respectively.
+
+Parameters to this method may be annotated with `@ShesmuParam` to give a
+description of the parameter.
+
+Since Shesmu and Java types are not entirely compatible, Shesmu will attempt to
+infer the Shesmu type from the Java type. If this isn't possible, both
+`@ShesmuMethod` and `@ShesmuParam` contain a `type` property which is the
+Shesmu type descriptor of the return type or parameter type.
+
+##### File-Backed Arbitrary Implementation
+For file-backed arbitrary implementations, functions can be created through the
+`UserDefiner.defineFunction` methods, which can take a instance of a function
+interface. The `VariadicFunction` interface takes an arbitrary number of
+arguments and the implementer of this interface is responsible for making sure
+the parameter types of the implementation match the parameter types of the
+definition.
+
+##### Manual Implementation
+When directly implementing the `DefinitionRepository` interface, create
+instances of `FunctionDefinition`.  There are a few ways methods to generate a
+`FunctionDefinition`:
 
 1. Create a class that implements `FunctionDefinition` and write arbitrary
-bytecode.
+bytecode. For example [`str_len` in `StandardDefinitions`](shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/core/StandardDefinitions.java).
 1. Use `FunctionDefinition.staticMethod` to create a binding for a public
 static method in a class.
 1. Use `RuntimeBinding` to create a binding for a method in an instance of a
```

Comment:
Is it worth having an explanation of what RunTimeBinding is? I notice there was one, in the deleted section headed Run Time Binding. Is that section now out of date? If not, is it a good idea to reinstate it?

Reply:
I think I should delete this advice. When each plugin was doing its own thing, they were using this class directly. Now it's basically a utility of the `FileBackMatchedDefinitionRepository`, so it shouldn't be mentioned.

Reply:
OK, makes sense.